### PR TITLE
[inductor][decomp] Add aten._unsafe_index_put for unchecked indexing

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102000
* __->__ #101812

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -829,6 +829,11 @@
   values: grad.index(indices)
   result: self_t.index_put(indices, values_t, accumulate)
 
+- name: _unsafe_index_put(Tensor self, Tensor?[] indices, Tensor values, bool accumulate=False) -> Tensor
```
Comment (lezcano) [inline tools/autograd/derivatives.yaml:832]: You should also use `_unsafe_index` in the backwards of `unsafe_index`

**Review by lezcano — APPROVED** [review]
PAM! Stamped.

**Comment by osalpekar** [comment]
This causes a large number of `test_nn_cuda` tests to crash internally. More details can be found at [D46093942](https://www.internalfb.com/diff/D46093942). cc @soulitzer. It might be useful to import this before trying to merge again, to ensure there are no breakages.

**Comment by peterbell10** [comment]
@osalpekar can you share any details for non-meta people?

**Comment by soulitzer** [comment]
@peterbell10 I think the internal failures are unrelated, trying the merge again

**Comment by ngimel** [comment]
The error doesn't look related
```
test_BCELoss_cuda_fp32 (unittest.loader._FailedTest) ... ERROR

======================================================================
ERROR: test_BCELoss_cuda_fp32 (unittest.loader._FailedTest)
----------------------------------------------------------------------
AttributeError: type object 'TestNN' has no attribute 'test_BCELoss_cuda_fp32'
```
(and similar errors for other tests)

**Pull request closed without merging**